From 866a56b6ce2d5537b676f7733e722f576096a065 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Tue, 25 Mar 2025 16:09:24 +0800
Subject: [PATCH 093/115] mtk: mac80211: fixup sanity check in
 ieee80211_build_hdr()

This might be a typo from upstream.

There were two conditions in the if-clause:
1. pre_conf_link_id != link_id
2. link_id != IEEE80211_LINK_UNSPECIFIED
and if the two conditions are true at the same time, the frame is
dropped.

The link_id is originally set to pre_conf_link_id and might be changed
to sta->deflink.link_id if sdata is an AP interface. In other word, the
only case that link_id is not equal to pre_conf_link is link_id becomes
sta->deflink.link_id, which is nerver IEEE80211_LINK_UNSPECIFIED.
In sammary, if the condition 1 is satisfied, the condition 2 must also
be satisfied and therefore becomes redundant.

The condition 2 might be a check on pre_conf_link_id not being
IEEE80211_LINK_UNSPECIFIED to prevent dropping the frame when then
link id is not specified originally.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 net/mac80211/tx.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/net/mac80211/tx.c b/net/mac80211/tx.c
index 5cb55744..751fa65a 100644
--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -2981,7 +2981,7 @@ static struct sk_buff *ieee80211_build_hdr(struct ieee80211_sub_if_data *sdata,
 						IEEE80211_TX_CTRL_MLO_LINK);
 
 		if (pre_conf_link_id != link_id &&
-		    link_id != IEEE80211_LINK_UNSPECIFIED) {
+		    pre_conf_link_id != IEEE80211_LINK_UNSPECIFIED) {
 #ifdef CPTCFG_MAC80211_VERBOSE_DEBUG
 			net_info_ratelimited("%s: dropped frame to %pM with bad link ID request (%d vs. %d)\n",
 					     sdata->name, hdr.addr1,
-- 
2.45.2

